modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/index/sorted/SortedIndexTree.java

```diff
@@ -67,18 +74,29 @@ public SortedIndexTree(
             AtomicLong globalRmvId,
             long metaPageId,
             @Nullable ReuseList reuseList,
-            boolean initNew,
-            BinaryTupleComparator binaryTupleComparator
+            SortedIndexDescriptor indexDescriptor,
+            boolean initNew
     ) throws IgniteInternalCheckedException {
         super("SortedIndexTree_" + grpId, grpId, grpName, partId, pageMem, lockLsnr, globalRmvId, metaPageId, reuseList);
 
-        setIos(SortedIndexTreeInnerIo.VERSIONS, SortedIndexTreeLeafIo.VERSIONS, SortedIndexTreeMetaIo.VERSIONS);
+        inlineSize = initNew ? InlineUtils.binaryTupleInlineSize(pageSize(), ITEM_SIZE_WITHOUT_COLUMNS, indexDescriptor)
+                : readInlineSizeFromMetaIo();
+
+        setIos(
+                SortedIndexTreeInnerIo.VERSIONS.get(inlineSize),
+                SortedIndexTreeLeafIo.VERSIONS.get(inlineSize),
+                SortedIndexTreeMetaIo.VERSIONS
+        );
 
         dataPageReader = new DataPageReader(pageMem, grpId, statisticsHolder());
 
-        this.binaryTupleComparator = binaryTupleComparator;
+        binaryTupleComparator = new BinaryTupleComparator(indexDescriptor);
 
         initTree(initNew);
+
+        if (initNew) {
+            writeInlineSizeToMetaIo(inlineSize);
```

Comment:
what if we fail for some reason before we had the chance to write the size to Meta IO?

Reply:
We will not save anything since the creation of indexes will be inside **MvPartitionStorage#runConsistently** i.e. checkpoint lock.